File: androidApp/src/main/java/fr/northborders/notesappkmm/android/note_list/NoteListViewModel.kt
package fr.northborders.notesappkmm.android.note_list

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import fr.northborders.notesappkmm.domain.note.Note
import fr.northborders.notesappkmm.domain.note.NoteDataSource
import fr.northborders.notesappkmm.domain.note.SearchNotes
import fr.northborders.notesappkmm.domain.time.DateTimeUtil
import fr.northborders.notesappkmm.presentation.RedOrangeHex
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class NoteListViewModel @Inject constructor(
    private val noteDataSource: NoteDataSource,
    private val savedStateHandle: SavedStateHandle // used to help restore state after process death
): ViewModel() {

    private val searchNotes = SearchNotes() // TODO should this be injected?

    private val notes = savedStateHandle.getStateFlow("notes", emptyList<Note>())
    private val searchText = savedStateHandle.getStateFlow("searchText", "")
    private val isSearchActive = savedStateHandle.getStateFlow("isSearchActive", false)

    val state = combine(notes, searchText, isSearchActive) { notes, searchText, isSearchActive ->
        NoteListState(
            notes = searchNotes.execute(notes, searchText),
            searchText = searchText,
            isSearchActive = isSearchActive
        )
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), NoteListState()) // on screen rotation we wouldn't resubscribe, only if you leave the screen for more than 5s

    // FIXME this is just for testing
//    init {
//        viewModelScope.launch {
//            (0 .. 10).forEach {
//                noteDataSource.insertNote(
//                    Note(
//                        id = it.toLong(),
//                        title = "Note$it",
//                        content = "Content$it",
//                        colorHex = RedOrangeHex,
//                        created = DateTimeUtil.now()
//                    )
//                )
//            }
//        }
//    }

    fun loadNotes() {
        viewModelScope.launch {
            savedStateHandle["notes"] = noteDataSource.getAllNotes()
        }
    }

    fun onSearchTextChange(text: String) {
        savedStateHandle["searchText"] = text
    }

    fun onToggleSearch() {
        savedStateHandle["isSearchActive"] = !isSearchActive.value
        if (!isSearchActive.value) {
            savedStateHandle["searchText"] = ""
        }
    }

    fun deleteNoteById(id: Long) {
        viewModelScope.launch {
            noteDataSource.deleteNoteById(id)
            loadNotes()
        }
    }
}

File: iosApp/iosApp/notes_list/NoteListViewModel.swift
//
//  NoteListViewModel.swift
//  iosApp
//
//  Created by Thibault Guégan on 13/10/2022.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation
import shared

extension NoteListScreen {
    @MainActor class NoteListViewModel: ObservableObject {
        private var noteDataSource: NoteDataSource? = nil
        
        private let searchNotes = SearchNotes()
        
        private var notes = [Note]()
        @Published private(set) var filteredNotes = [Note]()
        @Published var searchText = "" {
            didSet {
                self.filteredNotes = searchNotes.execute(notes: notes, query: searchText)
            }
        }
        @Published private(set) var isSearchActive = false
        
        init(noteDataSource: NoteDataSource? = nil) {
            self.noteDataSource = noteDataSource
            
            
        }
        
        func setNoteDataSource(noteDataSource: NoteDataSource) {
            self.noteDataSource = noteDataSource
            
//            noteDataSource.insertNote(note: Note(id: nil, title: "Title", content: "Content", colorHex: 0xFF456, created: DateTimeUtil().now()), completionHandler: { error in
//
//            })
        }
        
        func loadNotes() {
            noteDataSource?.getAllNotes(completionHandler: {notes, error in
                self.notes = notes ?? []
                self.filteredNotes = self.notes
            })
        }
        
        func deleteNoteById(id: Int64?) {
            if id != nil {
                noteDataSource?.deleteNoteById(id: id!, completionHandler: { error in
                    self.loadNotes()
                })
            }
        }
        
        func toggleSearchActive() {
            isSearchActive = !isSearchActive
            if !isSearchActive {
                searchText = ""
            }
        }
     }
}
